pyspecProcScripts/first_level/fake_data.py

```diff
@@ -0,0 +1,95 @@
+from .fl_dummy import fl_dummy
+from sympy import Symbol as sympy_symbol
+from sympy import lambdify
+from pyspecdata import nddata, ndshape
+import numpy as np
+from numpy.random import normal
+def fake_data(
+        expression,
+        axis_coords,
+        signal_pathway,
+        direct = 't2',
+        SD_sigma = [0.05,0.003],
+        SD_amp = [1,5],
+        scale=100,
+        fake_data_noise_std = 1.0,
+        fl=fl_dummy):
+    """Generate fake data subject to noise and frequency variation.
+    Parameters
+    ==========
+    expression: sympy expression
+        Gives the functional form of the data.
+    axis_coords: OrderedDict
+        Gives nddata objects providing all the axis coordinates.
+        **Very importantly**, these must be listed in the loop nesting order
+        (outside in) in which they occur in the pulse program,
+        or the frequency drift will not be modeled correctly.
+
+        To enable *simulating echo-like data*, you can specify a direct axis
+        that starts at a negative number.
+        If you do this, the beginning of the axis will be re-set to 0 before returning.
+    SD_sigma: list of floats
+        Gives the Gaussian σ for the spectral density of the
+        frequency variation.
+        Typically, there are more than one gaussian terms used to
+        create the spectral density.
+
+        A small σ gives rise to slow fluctuations while a large σ
+        gives rise to fast fluctuations.
+        The units of σ are in cycles per scan.
+    SD_amp: list of floats
+        Amplitudes associated with SD_sigma terms -- must be the
+        same length.
+    signal_pathway: dict
+        Gives the signal pathway, with keys being phase cycling dimensions, and
+        values being the corresponding Δp.
+    scale: float (default 100) 
+        amplitude of frequency variation
+    """
+    mysymbols = expression.atoms(sympy_symbol)
+    missing = (
+            set(str(j) for j in mysymbols)
+            - set(axis_coords.keys()))
+    assert len(missing) == 0, "all non-phase cycling symbols in your expression must have matching axis coordinates in the dictionary -- you are missing %s!" % str(missing)
+    thefunction = lambdify(mysymbols, expression, 'numpy')
+    clean_data = thefunction(*tuple(axis_coords[str(j)] for j in mysymbols))
```

Comment:
does the order of the axis_coords dictionary matter?

Reply:
Yes, it's explained in the docstring